### PR TITLE
docs: Describe workspace slot level caps

### PR DESCRIPTION
**Pull request by tara-det-ai**
#9180

[preview docs](https://determined-ai-docs.s3.us-west-2.amazonaws.com/previews/c51c93d1bda68cea8c8944dc5bdfae87/setup-cluster/k8s/resource-caps.html)

## Checklist

- [ ] Changes have been manually QA'd
- [ ] New features have been approved by the corresponding PM
- [ ] User-facing API changes have the "User-facing API Change" label
- [ ] Release notes have been added as a separate file under `docs/release-notes/`
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses have been included for new code which was copied and/or modified from any external code

[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+quotas using either the WebUI or CLI.
+
+*******
+ WebUI
+*******
+
+#. As an admin user, either create a new workspace or edit an existing workspace.
+
+#. In the "Namespace Bindings" section, type a name for the namespace or leave it blank. If left
+   blank, the workspace will be bound to the default namespace configured in the Master
```
Comment (hkumar92) [inline docs/setup-cluster/k8s/resource-caps.rst:22]: default namespace **for that K8s cluster**?

Comment (salonig23) [inline docs/setup-cluster/k8s/resource-caps.rst:22]: if the `default _namespace` field is not set in the master config it will be bound to the "default" namespace.

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 2]
```diff
@@ -0,0 +1,100 @@
+.. _k8s-resource-caps:
+
+#####################################
+ Manage Workspace Namespace Bindings
+#####################################
+
+.. note::
+
+   Managing resource quotas in Determined requires the Determined Enterprise Edition.
```
Comment (salonig23) [inline docs/setup-cluster/k8s/resource-caps.rst:9]: Auto creating namespace is also an EE feature

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+ WebUI
+*******
+
+#. As an admin user, either create a new workspace or edit an existing workspace.
+
+#. In the "Namespace Bindings" section, type a name for the namespace or leave it blank. If left
+   blank, the workspace will be bound to the default namespace configured in the Master
+   Configuration YAML file.
+
+#. Toggle the "Auto Create Namespace" option on or off. When enabled, you can edit the resource
```
Comment (salonig23) [inline docs/setup-cluster/k8s/resource-caps.rst:25]: Lets add that when enabled, determined will auto-create a namespace in the cluster

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+   quota directly in Determined.
+
+#. Enter a resource quota (e.g., 5). This quota is editable only when "Auto Create Namespace" is on.
+   If the quota is set in Kubernetes, it will appear here but won't be editable.
+
+When completed, save your changes to apply the namespace bindings and resource quotas.
```
Comment (hkumar92) [inline docs/setup-cluster/k8s/resource-caps.rst:31]: Do we want to say what the quota actually does? Like "any submitted workloads which would cause workspace resource use to exceed the defined cap will stay pending"

**Review by hkumar92 — CHANGES_REQUESTED** [review]

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+*****
+
+You can use the :ref:`command-line interface (CLI) <cli-ug>` to set one workspace namespace binding
+at a time. To get help for workspace bindings, run the following command:
+
+.. code:: bash
+
+   det w bindings -h
+
+Set a Namespace Binding
```
Comment (salonig23) [inline docs/setup-cluster/k8s/resource-caps.rst:50]: We also have a command to auto create namespaces:
`det w set <workspace-id> --cluster-name <cluste-name> --auto-create-namespace`
To autocreate namespaces for all clusters:
`det w set <workspace-id> --auto-create-namespace-all-clusters`

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+=======================
+
+To bind a workspace to a namespace for a particular cluster, use the following command:
+
+.. code:: bash
+
+   det w bindings set <workspace-id> --namespace <namespace-name>
```
Comment (salonig23) [inline docs/setup-cluster/k8s/resource-caps.rst:57]: to set bindings for a particular cluster, they will have to add the `--cluster-name <cluster-name>` arg 
This field is optional in Single RM because there will only be 1 cluster

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+Note: An error will be thrown if you try to delete a default binding.
+
+List Namespace Bindings
+=======================
+
+To list all namespace bindings, use:
+
+.. code:: bash
+
+   det w bindings list
```
Comment (salonig23) [inline docs/setup-cluster/k8s/resource-caps.rst:100]: We don't have a command to list all bindings from all workspaces. We can list bindings for a particular workspace by:
`det w bindings list <workspace-name>`

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+When completed, save your changes to apply the namespace bindings and resource quotas.
+
+Configure Additional Resource Managers
+======================================
+
+For each additional resource manager, configure the resource quota as needed. Each workspace in a
+cluster automatically has a default binding to their resource manager's namespace.
+
+*****
+ CLI
```
Comment (salonig23) [inline docs/setup-cluster/k8s/resource-caps.rst:40]: We can also add bindings and set quotas during workspace creation:
`det w create <workspace-name> --cluster-name <cluster-name> --namespace <namespace-name> --resource-quota <resource-quota>`
arguments such as `auto-create-namespace` and `auto-create-namespace-for-all` are also valid

File: docs/setup-cluster/k8s/resource-caps.rst (hunk 1 above)
Comment (tara-det-ai) [reply docs/setup-cluster/k8s/resource-caps.rst:22]: what is the correct wording to apply?

File: docs/setup-cluster/k8s/resource-caps.rst (hunk 4 above)
Comment (tara-det-ai) [reply docs/setup-cluster/k8s/resource-caps.rst:31]: Any submitted workloads that would cause workspace resource use to exceed the defined quota would stay pending (until the slot becomes available, e.g., when user kills a notebook that was using the slot).

File: docs/setup-cluster/k8s/resource-caps.rst (hunk 6 above)
Comment (tara-det-ai) [reply docs/setup-cluster/k8s/resource-caps.rst:57]: the system needs to know the name of the cluster. This field is optional with a single resource manager scenario (because there is only one cluster)

File: docs/setup-cluster/k8s/resource-caps.rst (hunk 1 above)
Comment (tara-det-ai) [reply docs/setup-cluster/k8s/resource-caps.rst:22]: ...to the default namespace configured for that k8s cluster in the master...

Comment (tara-det-ai) [reply docs/setup-cluster/k8s/resource-caps.rst:22]: edited as discussed

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 9]
```diff
@@ -6,7 +6,8 @@
 
 .. note::
 
-   Managing resource quotas in Determined requires the Determined Enterprise Edition.
+   Auto creating namespaces and managing resource quotas in Determined require the Determined
+   Enterprise Edition.
 
 Determined makes it easier to manage resource quotas by handling them directly, so you don't need to
 manage them separately in Kubernetes. You can manage workspace namespace bindings and resource
```
Comment (amandavialva01) [inline docs/setup-cluster/k8s/resource-caps.rst:13]: I think we should note that users can only manage resource quotas for auto-created namespaces.

Maybe `Determined makes it easier to manage resource quotas for auto-created namespaces by handling them directly, so you don't need to manage them separately in Kubernetes. You can set workspace-namespace bindings and resource quotas using either the WebUI or CLI.`

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 10]
```diff
@@ -18,23 +19,28 @@ quotas using either the WebUI or CLI.
 
 #. As an admin user, either create a new workspace or edit an existing workspace.
 
-#. In the "Namespace Bindings" section, type a name for the namespace or leave it blank. If left
-   blank, the workspace will be bound to the default namespace configured in the Master
-   Configuration YAML file.
+#. In the "Namespace Bindings" section, type a name for the namespace or leave it blank. The
+   behavior depends on the master configuration. For example, if the ``default_namespace`` is
+   defined, leaving the namespace blank binds the workspace to the "default" namespace (even if it
```
Comment (amandavialva01) [inline docs/setup-cluster/k8s/resource-caps.rst:24]: ...leaving the namespace field blank...

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 11]
```diff
@@ -18,23 +19,28 @@ quotas using either the WebUI or CLI.
...
+   is not set) configured for that Kubernetes cluster. If a different namespace name is provided,
+   the workspace is bound to that namespace.
```
Comment (amandavialva01) [inline docs/setup-cluster/k8s/resource-caps.rst:26]: In the "Namespace Bindings" section, enter a name for the namespace or leave it blank. If you specify a namespace in the namespace field, the workspace is bound to that namespace. If the field is left blank, the workspace is bound to the namespace specified in the ``resource_manager.default_namespace`` section of your master configuration when this configuration field is set, and is otherwise bound to the Kubernetes namespace, "default".

File: docs/setup-cluster/k8s/resource-caps.rst (hunk 9 above)
Comment (amandavialva01) [reply docs/setup-cluster/k8s/resource-caps.rst:13]: Since this is mentioned later on in the WebUI seciton, maybe we don't need to say it at the beginning? Not sure if it's worth specifying that detail here

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+
+#. Toggle the "Auto Create Namespace" option on or off. When enabled, the system automatically
+   creates a namespace in the cluster, allowing you to edit the resource quota directly in
+   Determined.
+
+#. Enter a resource quota (e.g., 5). This quota is editable only when "Auto Create Namespace" is on.
+   If the quota is set in Kubernetes, it will appear here but won't be editable.
```
Comment (amandavialva01) [inline docs/setup-cluster/k8s/resource-caps.rst:33]: Technically, even if someone edits a Determined-created quota in Kubernetes, they can still come back and set it again in Determined. Maybe we should reword this to say "If the quota for a namespace is created in Kubernetes, it will appear here but won't be editable".

We should also probably note that we display the lowest quota on the namespace, so if that's a Determined-created quota, the user can edit it in the WebUI and if the lowest quota was created in Kubernetes, then it won't be editable.

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+
+When completed, save your changes to apply the namespace bindings and resource quotas. Any submitted
+workloads that would cause workspace resource use to exceed the defined quota will stay pending
+until resources are available.
```
Comment (amandavialva01) [inline docs/setup-cluster/k8s/resource-caps.rst:37]: "until either resources are available or the quota is raised"? 
Not sure if we want to encourage frequently changing the quota though

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+
+Configure Additional Resource Managers
+======================================
+
+For each additional resource manager, configure the resource quota as needed. Each workspace in a
+cluster automatically has a default binding to its resource manager's namespace.
```
Comment (amandavialva01) [inline docs/setup-cluster/k8s/resource-caps.rst:43]: ...to its resource manager's configured or default namespace.

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+
+*****
+ CLI
+*****
+
+You can use the :ref:`command-line interface (CLI) <cli-ug>` to set one workspace namespace binding
+at a time. To get help for workspace bindings, run the following command:
```
Comment (amandavialva01) [inline docs/setup-cluster/k8s/resource-caps.rst:50]: ...To get help for setting workspace namespace bindings...

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+   det w bindings -h
+
+Create a Workspace with Bindings and Quotas
+===========================================
+
+You can add bindings and set quotas during workspace creation:
+
+.. code:: bash
+
+   det w create <workspace-name> --cluster-name <cluster-name> --namespace <namespace-name> --resource-quota <resource-quota>
```
Comment (amandavialva01) [inline docs/setup-cluster/k8s/resource-caps.rst:63]: Should we mention that `cluster-name` should be the ``resource_manager.cluster_name`` specified in master config? Or are we assuming users will know this intuitively?

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+
+Additional arguments such as ``--auto-create-namespace`` and ``--auto-create-namespace-for-all`` are
```
Comment (amandavialva01) [inline docs/setup-cluster/k8s/resource-caps.rst:65]: ...``--auto-create-namespace`` and ``--auto-create-namespace-all-clusters``...

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+also valid.
+
+Auto Create Namespaces
+======================
+
+To auto create a namespace for a specific cluster:
+
+.. code:: bash
+
+   det w set <workspace-id> --cluster-name <cluster-name> --auto-create-namespace
```
Comment (amandavialva01) [inline docs/setup-cluster/k8s/resource-caps.rst:75]: `det w bindings set <workspace-id> --cluster-name <cluster-name> --auto-create-namespace`

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+
+To auto create namespaces for all clusters:
+
+.. code:: bash
+
+   det w set <workspace-id> --auto-create-namespace-all-clusters
```
Comment (amandavialva01) [inline docs/setup-cluster/k8s/resource-caps.rst:81]: `det w bindings set <workspace-id> --auto-create-namespace-all-clusters`

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+
+   det w create <workspace-name> --cluster-name <cluster-name> --namespace <namespace-name> --resource-quota <resource-quota>
+
+Additional arguments such as ``--auto-create-namespace`` and ``--auto-create-namespace-for-all`` are
+also valid.
+
+Auto Create Namespaces
+======================
+
+To auto create a namespace for a specific cluster:
```
Comment (amandavialva01) [inline docs/setup-cluster/k8s/resource-caps.rst:71]: To bind a workspace to an auto-created namespace for a specific cluster:

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+Set a Namespace Binding
+=======================
+
+To bind a workspace to a namespace for a particular cluster, use the following command:
+
+.. code:: bash
+
+   det w bindings set <workspace-id> --cluster-name <cluster-name> --namespace <namespace-name>
+
+For a cluster with a single resource manager, the ``cluster-name`` is optional.
```
Comment (amandavialva01) [inline docs/setup-cluster/k8s/resource-caps.rst:92]: For a Determined cluster/ deployment with a single resource manager...
So that we distinguish the determined cluster with the Kubernetes cluster? 🙈😂

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 22]
```diff
@@ -0,0 +1,135 @@
...
+To auto create namespaces for all clusters:
+
+.. code:: bash
+
+   det w set <workspace-id> --auto-create-namespace-all-clusters
+
+Set a Namespace Binding
+=======================
+
+To bind a workspace to a namespace for a particular cluster, use the following command:
```
Comment (amandavialva01) [inline docs/setup-cluster/k8s/resource-caps.rst:86]: To bind a workspace to an already existing namespace for a particular cluster...

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 23]
```diff
@@ -0,0 +1,135 @@
...
+Example:
+
+.. code:: bash
+
+   det w bindings set ws2 --namespace ws2-899f-3
+
+Set a Resource Quota
+====================
+
+To set the resource quota for a workspace in a specific cluster, use:
```
Comment (amandavialva01) [inline docs/setup-cluster/k8s/resource-caps.rst:103]: To set the resource quota on a workspace for a specific cluster, use:

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 24]
```diff
@@ -0,0 +1,135 @@
...
+
+   det w resource-quota set <workspace-id> <quota> --cluster-name <cluster-name>
+
+Example:
+
+.. code:: bash
+
+   det w resource-quota set ws2 5 --cluster-name c1
+
+If you set resource quotas in Kubernetes, they will display in Determined but will not be editable.
```
Comment (amandavialva01) [inline docs/setup-cluster/k8s/resource-caps.rst:115]: Same comment from [above](https://github.com/determined-ai/determined/pull/9687/files#r1688210773) applies here I think

**Review by amandavialva01 — COMMENTED** [review]
This looks great, thanks for thorough documentation of workspace slot caps and detailed guides for setting them in the WebUI and CLI! 
I made a few suggestions, and I think it's worth noting somewhere in these docs that the resource quotas we set (and display in the webUI) in Determined specifically apply to Kubernetes GPU request limits (rather than other resource quotas like CPU or memory request limits). So in Determined, users can cap the GPU requests placed on a given workspace with the slot caps project, but other k8s resources available through a cluster's service provider are not managed with MLDE/ Determined.

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 25]
```diff
@@ -0,0 +1,135 @@
...
+#####################################
+ Manage Workspace Namespace Bindings
+#####################################
+
+.. note::
+
+   Auto creating namespaces and managing resource quotas in Determined require the Determined
+   Enterprise Edition.
+
+Determined makes it easier to manage resource quotas by handling them directly, so you don't need to
```
Comment (salonig23) [inline docs/setup-cluster/k8s/resource-caps.rst:12]: @hkumar92 should we warn the users about [this](https://hpe-aiatscale.slack.com/archives/C06RZSV56FM/p1721077493638219) case here?

File: docs/setup-cluster/k8s/resource-caps.rst (hunk 18 above)
Comment (tara-det-ai) [reply docs/setup-cluster/k8s/resource-caps.rst:75]: good catch

File: docs/setup-cluster/k8s/resource-caps.rst (hunk 24 above)
Comment (tara-det-ai) [reply docs/setup-cluster/k8s/resource-caps.rst:115]: i'll remove this instance to make it easier to maintain and harder for human error to show up in the docs

File: docs/setup-cluster/k8s/resource-caps.rst (hunk 25 above)
Comment (hkumar92) [reply docs/setup-cluster/k8s/resource-caps.rst:12]: Yes we should let them know especially of potential footguns we've tested and aware of

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 26]
```diff
@@ -0,0 +1,135 @@
+.. _k8s-resource-caps:
+
+#####################################
+ Manage Workspace Namespace Bindings
```
Comment (hkumar92) [inline docs/setup-cluster/k8s/resource-caps.rst:4]: Should we call it workspace - namespace bindings? Or Workspace <> Namespace bindings or something?

File: docs/setup-cluster/k8s/resource-caps.rst (hunk 16 above)
Comment (tara-det-ai) [reply docs/setup-cluster/k8s/resource-caps.rst:63]: added

**Review by hkumar92 — CHANGES_REQUESTED** [review]

File: docs/setup-cluster/k8s/resource-caps.rst (hunk 24 above)
Comment (tara-det-ai) [reply docs/setup-cluster/k8s/resource-caps.rst:115]: i'll leave the single instance above and remove it from the cli section so there aren't two instances to maintain

File: docs/setup-cluster/k8s/resource-caps.rst (hunk 25 above)
Comment (salonig23) [reply docs/setup-cluster/k8s/resource-caps.rst:12]: @tara-det-ai can we add a warning that if a job is pending due to lack of resources (not resource quota), and they decide to change the resource quota, the job will get scheduled despite any restricting resource quota

Comment (tara-det-ai) [reply docs/setup-cluster/k8s/resource-caps.rst]: thank you for the summarized description of the use case. i've added an "important" admonition.

File: docs/setup-cluster/k8s/resource-caps.rst (hunk 26 above)
Comment (tara-det-ai) [reply docs/setup-cluster/k8s/resource-caps.rst:4]: yes it makes sense to add a hyphen

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 27]
```diff
@@ -0,0 +1,173 @@
...
+-  Description: Use this endpoint to fetch the workspace IDs of workspaces that have at least one
+   default binding.
+-  Usage: This can help identify which workspaces need namespace bindings to be auto-created.
+
+Bulk Auto-Create Namespace Bindings
+===================================
+
+-  Endpoint: ``/api/v1/namespace-bindings/bulk-auto-create``
+-  Description: Use this endpoint to auto-create namespace bindings for all specified workspaces.
+-  Details: Pass the workspace IDs fetched from the previous endpoint into this endpoint. It will
```
Comment (salonig23) [inline docs/setup-cluster/k8s/resource-caps.rst:170]: For each workspace, it will ...

**Review by salonig23 — APPROVED** [review]
Apart from 1 comment, everything looks great to me. Thank you Tara for being so thorough and helping us write this doc!

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 28]
```diff
@@ -0,0 +1,177 @@
...
+
+Create a Workspace with Bindings and Quotas
+===========================================
+
+You can add bindings and set quotas during workspace creation using the following command where
+``cluster-name`` is the ``resource_manager.cluster_name`` specified in the master configuration:
+
+.. code:: bash
+
+   det w create <workspace-name> --cluster-name <cluster-name> --namespace <namespace-name> --resource-quota <resource-quota>
```
Comment (amandavialva01) [inline docs/setup-cluster/k8s/resource-caps.rst:79]: `det w create <workspace-name> --cluster-name <cluster-name> --namespace <namespace-name>`

Comment (amandavialva01) [reply docs/setup-cluster/k8s/resource-caps.rst:79]: We can't set the resource quota on workspaces bound to namespaces that weren't auto-created by Determined

File: docs/setup-cluster/k8s/resource-caps.rst [hunk 29]
```diff
@@ -0,0 +1,177 @@
...
+   det w bindings set <workspace-id> --auto-create-namespace-all-clusters
+
+Set a Namespace Binding
+=======================
+
+To bind a workspace to an existing namespace for a particular cluster, use the following command:
+
+.. code:: bash
+
+   det w bindings set <workspace-id> --cluster-name <cluster-name> --namespace <namespace-name>
```
Comment (amandavialva01) [inline docs/setup-cluster/k8s/resource-caps.rst:106]: Ref comment [here](https://github.com/determined-ai/determined/pull/9687/files#r1690263010)

File: docs/setup-cluster/k8s/resource-caps.rst (hunk 28 above)
Comment (tara-det-ai) [reply docs/setup-cluster/k8s/resource-caps.rst:79]: fix in #9720
